### PR TITLE
Add spans to hooks

### PR DESCRIPTION
**Pull request by marcus13371337**
Fixes:
https://github.com/autotelic/fastify-opentelemetry/issues/56

### CODE REVIEW

**Comment by marcus13371337** [comment]
@HW13  can you have a look? 🙏

File: index.js [hunk 1]
```diff
@@ -122,36 +127,52 @@ async function openTelemetryPlugin (fastify, opts = {}) {
...
-        return context.with(reqContext, routeHandler.bind(this, request, ...args))
+  function executeIfNotWrappedOrNotIgnoredRoute (handler) {
+    return async function (request, ...params) {
+      if (!shouldWrapRoute(request.routerPath) && !shouldIgnoreRoute(request.url, request.method)) {
+        return handler(request, ...params)
       }
     }
+  }
+
+  fastify.addHook('onRequest', onRequest)
```
Comment (HW13) [inline index.js:143]: Could we use the same pattern as `onResponse` and on `onError` here as well? This way we can ensure that our `onRequest` hook always get called first

```js
fastify.addHook('onRequest', executeIfNotWrappedOrNotIgnoredRoute(onRequest))

// ...

routeOpts.onRequest = [onRequest, ...normalizeHooks(routeOpts.onRequest)]
```

File: index.js [hunk 2]
```diff
@@ -122,36 +127,52 @@ async function openTelemetryPlugin (fastify, opts = {}) {
...
+          const reqContext = getContext(request)
+          return context.with(reqContext, handler.bind(this, request, ...args))
         }
       }
-    })
-  }
+
+      routeOpts.handler = wrapHandler(handler)
+
+      function normalizeHooks (hooks) {
```
Comment (HW13) [inline index.js:160]: Could we update this function to wrap each hook in `context`? Similar to your suggestion -> https://github.com/autotelic/fastify-opentelemetry/issues/56#issuecomment-1444841874

```js
function wrapHooks (hooks) {
  if (!hooks) {
    return []
  }

  if (Array.isArray(hooks)) {
    return hooks.map(wrapHandler)
  }

  return [wrapHandler(hooks)]
}

// Then just update remaining hook types...
routeOpts.onRequest = [onRequest, ...wrapHooks(routeOpts.onRequest)]
routeOpts.preParsing = wrapHooks(routeOpts.preParsing)
routeOpts.preValidation = wrapHooks(routeOpts.preValidation)
// ... etc.
```

File: index.js [hunk 3]
```diff
@@ -122,36 +127,52 @@ async function openTelemetryPlugin (fastify, opts = {}) {
...
+
+        if (Array.isArray(hooks)) {
+          return hooks
+        }
+
+        return [hooks]
+      }
+
+      routeOpts.onResponse = [...normalizeHooks(routeOpts.onResponse), onResponse]
+      routeOpts.onError = [...normalizeHooks(routeOpts.onError), onError]
```
Comment (HW13) [inline index.js:173]: Great idea! 🧠  This will ensure we don't end our span until after all other hooks have completed

File: example/index.js [hunk 4]
```diff
@@ -32,6 +32,21 @@ const start = async () => {
...
+  fastify.route({
+    url: '/with-hooks',
+    method: 'GET',
+    handler: async function routeHandler (request, reply) {
+      return 'Hello World With hooks!'
+    },
+    preValidation: function preValidation (request, reply, done) {
+      const { activeSpan } = request.openTelemetry()
+
+      activeSpan.addEvent('preValidation')
```
Comment (HW13) [inline example/index.js:44]: Could we have this example start and end a new span instead? Would like to show that spans from hooks are now automatically children of the plugin's span 😎

**Review by HW13 — CHANGES_REQUESTED** [review]
Apologies for the delay @marcus13371337!

Thanks for the PR, this looks great! Just have a few requests 🙏

**Comment by HW13** [comment]
#65 has implemented this functionality

**Pull request closed without merging**